packages/svelte/src/compiler/phases/utils.js

```diff
@@ -0,0 +1,81 @@
+/** @import { AST, Binding } from '#compiler' */
+/** @import { Scope } from './scope' */
+/** @import * as ESTree from 'estree' */
+
+import { walk } from 'zimmerframe';
+
+/**
+ * @param {ESTree.Expression} expr
+ */
+export function extract_identifiers(expr) {
+	/** @type {ESTree.Identifier[]} */
+	let nodes = [];
+
+	walk(
+		expr,
+		{},
+		{
+			Identifier(node, { path }) {
+				const parent = path.at(-1);
+				if (parent?.type !== 'MemberExpression' || parent.property !== node || parent.computed) {
+					nodes.push(node);
+				}
+			}
+		}
+	);
+
+	return nodes;
+}
+
```

Comment:
Walking is expensive, we want to do it as infrequently as possible. Luckily we don't need this function anyway, since the necessary information exists on `value.metadata.expression.dependencies`.

As a side-note, the check in the `Identifier` visitor is insufficient. If you want to know if an identifier is a _reference_ then you can use [`is-reference`](https://github.com/Rich-Harris/is-reference/)

```suggestion
```

Reply:
...although `value.metadata.expression.dependencies` excludes globals, which means that things like `{location}` get inlined incorrectly. We may need to treat globals as bindings